### PR TITLE
Add image information table to container details panel

### LINKED ISSUE #1800: Show which image a container is running
In order of importance:
- image name
- image label
- link to where image is from (e.g. dockerhub, quay.io URL)

Secondarily:
- size
- age
- other stuff

Ideally also directly on pod, replication controller, and deployment views too, since it should be the same on all of them.

Kubernetes views should highlight if pods in the same rc, deployment, etc. differ (they really never should, but if they _do_, then that's very interesting).

### PR DESCRIPTION
**Pull request by 2opremio**
Fixes #1800 

screen shot 2016-10-20 at 13 24 07

### CODE REVIEW

**Comment by jml** [comment]
Thanks!

I think we should re-order the headings: STATUS, IMAGE, DOCKER LABELS, ENVIRONMENT VARIABLES, PROCESSES.

**Comment by rade** [comment]
Can we shift the whole thing to the left, so less space is used by keys and more by values?

File: probe/docker/reporter.go [hunk 1]
```diff
@@ -237,16 +249,17 @@ func (r *Reporter) containerImageTopology() report.Topology {
...
-		node = node.AddTable(ImageLabelPrefix, image.Labels)
-
+		latests := map[string]string{
+			ImageID:          imageID,
+			ImageSize:        humanize.Bytes(uint64(image.Size)),
+			ImageVirtualSize: humanize.Bytes(uint64(image.VirtualSize)),
+		}
 		if len(image.RepoTags) > 0 {
-			node = node.WithLatests(map[string]string{ImageName: image.RepoTags[0]})
+			latests[ImageName] = image.RepoTags[0]
```
Comment (jml) [inline probe/docker/reporter.go]: How does this work? Is the first thing in `RepoTags` the latest?

**Comment by 2opremio** [comment]
> Can we shift the whole thing to the left, so less space is used by keys and more by values?

@davkal / @foot Wanna do the honors? :)

File: report/table.go [hunk 2]
```diff
@@ -17,8 +17,8 @@ const (
 	TruncationCountPrefix = "table_truncation_count_"
 )
 
-// AddTable appends arbirary key-value pairs to the Node, returning a new node.
-func (node Node) AddTable(prefix string, labels map[string]string) Node {
+// AddPrefixTable appends arbirary key-value pairs to the Node, returning a new node.
```
Comment (jml) [inline report/table.go]: Nit: "arbitrary"

File: report/table_test.go [hunk 3]
```diff
@@ -28,6 +28,29 @@ func TestTables(t *testing.T) {
...
+	nmd := report.MakeNodeWith("foo1", map[string]string{
+		"foo1key": "bar1",
+		"foo2key": "bar2",
+	})
+
+	template := report.TableTemplate{FixedRows: map[string]string{
+		"foo1key": "foo1",
+		"foo2key": "foo2",
+	},
+	}
```
Comment (jml) [inline report/table_test.go]: Kind of surprised this what gofmt likes. No action required.

**Comment by rade** [comment]
> I think we should re-order the headings: STATUS, IMAGE, DOCKER LABELS, ENVIRONMENT VARIABLES, PROCESSES. 

I am not convinced. 'Processes' is navigable (i.e. you can click on things and get new detail panels). It's weird to have to scroll in order to navigate.

**Comment by jml** [comment]
> I am not convinced. 'Processes' is navigable (i.e. you can click on things and get new detail panels). It's weird to have to scroll in order to navigate.

I just listed things in order of interest to me.

**Comment by 2opremio** [comment]
> I think we should re-order the headings: STATUS, IMAGE, DOCKER LABELS, ENVIRONMENT VARIABLES, PROCESSES.

I can definitely show the image info first, but  showing the processes at the end (which I am also doubtful about) requires a UI change (which will likely affect all details panels unless we want to start making exceptions).

IMAGE, DOCKER LABELS and ENVIRONMENT VARIABLES are normal tables and PROCESSES is a child table. The relative rendering order between tables and child tables is determined by the UI at https://github.com/weaveworks/scope/blob/master/client/app/scripts/components/node-details.js#L192 (kudos to @paulbellamy for finding out)

**Comment by davkal** [comment]
It's a bit weird showing all those details in the container details panel, but not in the container-image details panel. The current metadata section looks a bit barren:

screen shot 2016-10-20 at 22 48 57

**Comment by 2opremio** [comment]
> It's a bit weird showing all those details in the container details panel, but not in the container-image details panel. The current metadata section looks a bit barren:

I agree but that is part of #925. The nodes in the image topoly represent an image name (which potentially corresponds to multiple images). Doing this right involves rendering tables to include the IDs/Sizes/Tags of multiple images is a bit harder to achieve and is out of the scope of this PR.

> This also raises the question, shouldnt it be enough to be there?

I asked this to @jml and he claimed it's not the case (please see above)

> If it's so important as @jml claims, I suggest to put the first 3 items in the meta data section of the container

Uhm, that section is already crowded and requires expanding, that's why I went for a separate table (and added the extra table type required to make that possible)

**Comment by davkal** [comment]
> Uhm, that section is already crowded and requires expanding, that's why I went for a separate table (and added the extra table type required to make that possible)

jml's point was that currently shown priority (i.e. non-expanded) fields are useless to him. My suggestion was to add the newly important fields as priority fields and demote some current priority fields. That way it will look less cluttered/crowded and will require no scrolling.

**Comment by 2opremio** [comment]
I agree but that is part of #925. The nodes in the image to poly represent
an image name (which potentially corresponds to multiple images). That is a
bit harder to achieve and is out of the scope of this PR.

**Comment by 2opremio** [comment]
> My suggestion was to add the newly important fields as priority fields and demote some current priority fields

Yep, got it :) However I don't think the image details (e.g. image size) aren't the most important information about the container and we only display 3 fields in the image.

> That way it will look less cluttered/crowded and will require no scrolling.

It doesn't require scrolling on a Desktop screen (as per the screenshot attached).

**Comment by jml** [comment]
@2opremio asked me:

> [October 27th, 2016 6:23 AM] fons: @jml Can you unblock https://github.com/weaveworks/scope/pull/1942 by choosing the option which is better IYO?

After reading this carefully, I am not clear on what the alternatives are.

**Comment by 2opremio** [comment]
> After reading this carefully, I am not clear on what the alternatives are.

The alternatives are:
1. Leave the image information in its own separate table, as it is in the current PR. (my inclination)
2. Move the image information to the main metadata section (_Info_ section). (@davkal's inclination)

**Comment by jml** [comment]
Thanks for clarifying, @2opremio. 

tl;dr: I think we should:

* put the image name & label as one field "Image" in the top Info table
* either 
  * change Info to allow more fields as visible by default 
  * push container ID "below the fold" (i.e. only visible when Info expanded)
* put the rest of the image details in an image section at the bottom, as it is now, possibly repeating the image name & label stuff for context

Thoughts and a huge pile of questions below.

> we only display 3 fields in the metadata without expanding

This seems pretty arbitrary (we display more environment variables, for example). It also seems that the things we've chosen to be at the top are also fairly arbitrary. Who do you think needs the Container ID first thing? What would they use it for?

What does the command mean separately from the context of the image? How would it be useful to someone who doesn't know what the image is? How useful is it _really_ to show only the first 20 characters of the command line?

> IMAGE, DOCKER LABELS and ENVIRONMENT VARIABLES are normal tables and PROCESSES is a child table.

This seems like an implementation detail? I don't know the difference between these, and I suspect most users will be similar. If a layout is more helpful to users, wouldn't we change our internal structure to accommodate them, rather than expecting them to adapt to what's easy for us?

> However I don't think the image details (e.g. image size) are the most important information about the container

Image size is definitely not most important. It's basically an irrelevance almost all of the time.

However, image name & label are almost certainly the most important thing about the container. At least, they are to me.

**Comment by rade** [comment]
Let's not turn this PR into a complete redesign and re-architecting of the details panel. Modulo easy tweaks, let's get this merged provided we believe it improves on the status quo, which it does, doesn't it? And file a follow-up issue for improvements.

**Comment by 2opremio** [comment]
>  Let's not turn this PR into a complete redesign and re-architecting of the details panel. Modulo easy tweaks, let's get this merged provided we believe it improves on the status quo, which it does, doesn't it? And file a follow-up issue for improvements.

Agreed.

From the changes proposed by @jml I can easily do the following:

> * put the image name & label as one field "Image" in the top Info table
> * push container ID "below the fold" (i.e. only visible when Info expanded)
> * put the rest of the image details in an image section at the bottom, as it is now, possibly repeating the image name & label stuff for context
